torch/_higher_order_ops/scan.py

```diff
@@ -0,0 +1,339 @@
+# mypy: allow-untyped-defs
+import functools
+import itertools
+from typing import Callable, List
+
+import torch
+import torch._prims_common as utils
+import torch._subclasses.functional_tensor
+import torch.utils._pytree as pytree
+from torch._C import DispatchKey
+from torch._higher_order_ops.utils import (
+    _set_compilation_env,
+    autograd_not_implemented,
+    reenter_make_fx,
+    unique_graph_id,
+)
+from torch._ops import HigherOrderOperator
+from torch._subclasses.fake_tensor import FakeTensorMode
+from torch.fx.experimental.proxy_tensor import (
+    disable_proxy_modes_tracing,
+    ProxyTorchDispatchMode,
+    track_tensor_tree,
+)
+from torch.utils._python_dispatch import _get_current_dispatch_mode
+
+
+aten = torch._ops.ops.aten
+
+
+def wrap_combine_fn_flat(*args, combine_fn, spec, num_leaves):
+    assert len(args) == 2 * num_leaves
+    lhs = pytree.tree_unflatten(args[:num_leaves], spec)
+    rhs = pytree.tree_unflatten(args[num_leaves:], spec)
+    combined = combine_fn(lhs, rhs)
+    combined_leaves = pytree.tree_leaves(combined)
+    assert num_leaves == len(combined_leaves)
+    return combined_leaves
+
+
+def scan(
+    combine_fn: Callable[[pytree.PyTree, pytree.PyTree], pytree.PyTree],
+    input: pytree.PyTree,
+    dim: int,
+    reverse: bool = False,
+    init: pytree.PyTree = None,
+) -> torch.Tensor:
+    r"""
+    Performs an inclusive scan with a combine function.
+
+    .. warning::
+        `torch.scan` is a prototype feature in PyTorch. It currently
+        does not support autograd and you may run into miscompiles.
+        Read more about feature classification at:
+        https://pytorch.org/blog/pytorch-feature-classification-changes/#prototype
+
+    Args:
+        combine_fn (Callable): A binary callable with type ``(Tensor, Tensor) -> Tensor``,
+            or if input is a pytree ``(pytree, pytree) -> pytree``.
+            This function must be pure, i.e., no lifted arguments are supported at the moment.
+        input (torch.Tensor): The input tensor, or nested pytree of tensors.
+        dim (int): the dimension to scan over
+        reverse (bool): A boolean stating if the scan should be reversed with respect to ``dim``, default ``False``.
+        init (torch.Tensor): The inital scan carry, a tensor, or nested pytree of tensors that
+            represents the first output of the scan, default ``None``. The ``init`` is expected to have the
+            same pytree structure and shape as the output tensors of ``combine_fn``.
+            In case the ``init`` is ``None``, the first element of input is used as ``init``.
+
+
+    Example::
+
+        def add(x: torch.Tensor, y: torch.Tensor):
+            return x + y
+
+        # Usage without the usage of ``init``
+        cumsum = scan(add, x, dim)
+        # This produces the output
+        cumsum = [x0, add(x0, x1), add(x1, x2)]
```

Comment:
1. The output dtype should be a tensor, right? We can have torch.stack around the current output list to make this clearer.
2. the output should look like [x0, add(x0, x1), add(add(x0, x1), x2)], right? Right now, it's not a accumulative sum i think.
3. When init is not None, is init included as part of output? What's the shape of the output in the scanned dimension?

Reply:
Ad 1.) I corrected the docstring
Ad 2.) You are fully right, I corrected the docstring
Ad 3.) Init is included as part of the output. I tried to cover this in the example in the docstring. In the current implementation, the shape of init has to be the same shape as the output of the scan combine_fn. The shape of output for the scanned dimension is always 1.